### PR TITLE
Issue/434 Implement Interactives

### PR DESCRIPTION
**Pull request by hayleyavw**
Alrighty, get ready, this is it.

I present to you, THE INTERACTIVES PR!! * woooooooo * 

This PR implements interactives in the new django system, very exciting:
- Added an interactive application in order to implemented in-page, iframe and whole-page interactives
- Switched django system to be a bit more ``sass``y
- Updated a few pieces of infrastructure 
- Added a few tests
- Documentation for new interactives application
- Added the coding encryption chapter

Go through it as thoroughly as you can and make comments on everything terrible (and everything nice, this took a long time after all).
This PR is a big one and it dabbled in a lot of different areas, so not every issue you find will be fixed in this PR, at some point it will need to be merged and the remaining/less urgent issues will need to be fixed in a different PR.

### CODE REVIEW

**Comment by JackMorganNZ** [comment]
I believe this is set to merge into the wrong branch? I don't think I'm supposed to be checking 60,000 lines.

**Comment by hayleyavw** [comment]
@JackMorganNZ fixed. Sorry, been having this issue a lot lately where github keeps changing the first branch I set in a comparison.

File: csfieldguide/chapters/management/commands/_GlossaryTermsLoader.py [hunk 1]
```diff
@@ -20,16 +20,13 @@ def __init__(self, structure_file_path, glossary_directory_name, BASE_PATH):
...
         self.glossary_folder_path = os.path.join(self.BASE_PATH, glossary_directory_name)
         self.FILE_EXTENSION = ".md"
 
     @transaction.atomic
     def load(self):
         """Load the glossary content into the database."""
-        # glossary_slugs = set()
-
         for filename in listdir(self.glossary_folder_path):
             if filename.endswith(self.FILE_EXTENSION):
```
Comment (JackMorganNZ) [inline csfieldguide/chapters/management/commands/_GlossaryTermsLoader.py]: Files should require `.md` extension. There is no other files that have this option.

File: csfieldguide/config/settings/base.py [hunk 2]
```diff
@@ -158,11 +159,13 @@
 STATIC_ROOT = str(ROOT_DIR.path("staticfiles"))
 
 # See: https://docs.djangoproject.com/en/dev/ref/settings/#static-url
+BUILD_ROOT = os.path.join(str(ROOT_DIR.path("build")), "")
 STATIC_URL = "/staticfiles/"
 
 # See: https://docs.djangoproject.com/en/dev/ref/contrib/staticfiles/#std:setting-STATICFILES_DIRS
 STATICFILES_DIRS = [
     str(ROOT_DIR.path("static")),
+    BUILD_ROOT,
```
Comment (JackMorganNZ) [inline csfieldguide/config/settings/base.py]: If this is added, the previous line should be removed.

File: csfieldguide/interactives/management/commands/_InteractivesLoader.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        # Create interactive object and save to the db
+        new_interactive = Interactive(
+            slug=self.interactive_slug,
+            name=interactive_name,
+            template="interactives/{}.html".format(self.interactive_slug)
+        )
+        new_interactive.save()
+
+        self.log("Added interactive: {}".format(new_interactive.name))
```
Comment (JackMorganNZ) [inline csfieldguide/interactives/management/commands/_InteractivesLoader.py]: This uses different case to chapter log message. I would prefer the chapter log to match this one.

File: csfieldguide/interactives/management/commands/_InteractivesLoader.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                self.structure_file_path,
+                ["name"],
+                "Interactive"
+            )
+
+        # Create interactive object and save to the db
+        new_interactive = Interactive(
+            slug=self.interactive_slug,
+            name=interactive_name,
+            template="interactives/{}.html".format(self.interactive_slug)
```
Comment (JackMorganNZ) [inline csfieldguide/interactives/management/commands/_InteractivesLoader.py]: Further discussion is required regarding where the templates of interactives are stored. 90% of interactive data (JS/CSS/README/etc) are stored in a folder in the `static` folder while the `HTML` file is within the interactives *not* in a folder of its own. It seems inconsistent.

- Why aren't the templates within their own folder?
- Why aren't the static files and templates within the interactive's application folder?
- Is it possible for an interactive to have multiple HTML pages?

File: csfieldguide/interactives/management/commands/_InteractivesLoader.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+"""Custom loader for loading an interactive."""
```
Comment (JackMorganNZ) [inline csfieldguide/interactives/management/commands/_InteractivesLoader.py]: Would be better as:

> Custom loader for loading interactives.

File: csfieldguide/interactives/management/commands/loadinteractives.py [hunk 6]
```diff
@@ -1,30 +1,49 @@
...
     """Required command class for the custom Django loadinteractives command.
 
     Raises:
-        MissingRequiredFieldError: when no object can be found with the matching
-                attribute.
+        MissingRequiredFieldError: When a config (yaml) file is missing a
+            required field.
     """
 
     help = "Converts Markdown files listed in structure file and stores"
```
Comment (JackMorganNZ) [inline csfieldguide/interactives/management/commands/loadinteractives.py]: This is incorrect.

File: csfieldguide/interactives/management/commands/loadinteractives.py [hunk 7]
```diff
@@ -1,30 +1,49 @@
...
-        # Hardcoded for testing, TODO this should be in _InteractiveLoader.py
-        new_interactive = Interactive(
-            slug="sorting-algorithm-comparison",
-            template="interactives/sorting-algorithm-comparison.html"
-        )
-        new_interactive.save()
+        factory = LoaderFactory()
+
+        base_loader = BaseLoader()
+        BASE_PATH = "interactives/"
```
Comment (JackMorganNZ) [inline csfieldguide/interactives/management/commands/loadinteractives.py]: This should be stored in the settings file for overriding for tests.

File: csfieldguide/interactives/models.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
 """Models for the interactives application."""
 
 from django.db import models
+# from chapters.models import Chapter
```
Comment (JackMorganNZ) [inline csfieldguide/interactives/models.py]: This should be deleted not commented.

File: csfieldguide/static/interactives/caesar-cipher/README.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
+## Description
+
+This interactive encrypts and decrypts using the Caesar Cipher.
+
+## Usage
+
+To use this within the Field Guide it should be use an interactive tag (e.g. `{interactive name="caesar-cipher" type="iframe"}`).
+
+
```
Comment (JackMorganNZ) [inline csfieldguide/static/interactives/caesar-cipher/README.md]: Extra newline.

File: csfieldguide/static/interactives/caesar-cipher/README.md [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+## Development
+
+In order to make changes you will need nodejs, the source is written in CoffeeScript (don't edit the compiled JavaScript directly as it gets automatically generated).
```
Comment (JackMorganNZ) [inline csfieldguide/static/interactives/caesar-cipher/README.md]: Should be `Node.js`

File: csfieldguide/static/interactives/caesar-cipher/index.html [hunk 11]
```diff
@@ -0,0 +1,37 @@
+<!DOCTYPE html>
```
Comment (JackMorganNZ) [inline csfieldguide/static/interactives/caesar-cipher/index.html]: Why is this file here?

File: docs/source/developer/interactives.rst [hunk 12]
```diff
@@ -1,126 +1,163 @@
...
+- ``iframe``: This is an interactive that is embedded in a chapter by using an iframe.
+    This is used if the interactive is included multiple times in the chapter (to avoid conflicts in JS/CSS).
+- ``whole-page``: This in an interactive that is displayed on a new page, which is accessible via a button from the chapter.
 
-Interactive Types
+Adding Interactives
 ==============================================================================
 
-Each *interactive* is stored in it's own folder within the `interactives` folder. An author can add an interactive to the text using the [interactive tag](docs/guide-content.md#interactives), and the generator will include the interactive in one of the following ways (depending on the setting within the tag):
+Once you have developed your interactive (see ***insert link here***), it's time to add it to the chapter.
```
Comment (JackMorganNZ) [inline docs/source/developer/interactives.rst]: I presume `(see ***insert link here***)` should be replaced>

File: docs/source/developer/interactives.rst [hunk 13]
```diff
@@ -1,126 +1,163 @@
...
+This includes embedding the interactive in the chapter text see :ref:`writing-guide-interactive`.
 
-- `in-page` - The interactive is included by copying specific HTML/CSS/JavaScript (which has been flagged with a class) into the page. This is the preferred method for including an interactive on a page.
-- `whole-page` - Creates a link to the interactive displayed on a separate page. The link shows a thumbnail of the interactive with a text (the text is set using the `text` parameter).
-- `iframe` - The interactive is included in the page by embedding the interactive within an iframe. This is used if the interactive is included multiple times on the page (to avoid conflicts in JavaScript/CSS).
+.. note::
+  
+  When adding an interactive using the verto syntax, note that verto uses the term ``name``, this is actually referring to the ``slug`` of the interactive.
```
Comment (JackMorganNZ) [inline docs/source/developer/interactives.rst]: Should be "Verto" instead of "verto".

File: docs/source/developer/interactives.rst [hunk 14]
```diff
@@ -1,126 +1,163 @@
...
 
+- ``templates/interactives/<interactive-slug>.html`` - This is the HTML template for the interactive and it's name must match the slug of the interactive.
+ 
+.. note::
 
-JavaScript Files
+  When creating a new interactive, we suggest you copy the ``static/interactives/example-structure/`` directory and rename it to the slug for your interactive.
+  This directory has the basic structure already mapped out to get you started.
+
+  Similarly, within the ``templates/interactives/`` directory, there is a starter template for each of the interactive types (``iframe-template.html``, ``in-page-template.html`` and ``whole-page-template.html``).
```
Comment (JackMorganNZ) [inline docs/source/developer/interactives.rst]: I was very confused what these templates were until I read this comment (at the end of the PR). It would be wise to prefix all these with something like `example-` or `starter-`.

File: docs/source/developer/interactives.rst [hunk 15]
```diff
@@ -1,126 +1,163 @@
...
-- `third-party/jsencrypt.js` - [JSEncrypt](http://travistidwell.com/jsencrypt/) is a Javascript library to perform OpenSSL RSA Encryption, Decryption, and Key Generation.
-- `third-party/modernizr.js` - [Modernizr](https://modernizr.com/) is a small piece of JavaScript code that automatically detects the availability of next-generation web technologies in your user's browsers. Rather than blacklisting entire ranges of browsers based on “UA sniffing,” Modernizr uses feature detection to allow you to easily tailor your user's experiences based on the actual capabilities of their browser.
-- `third-party/nouislider.min.js` - Event handling for the [noUiSlider library](http://refreshless.com/nouislider/), used within many colour interactives.
-- `third-party/wNumb.js` - [wNumb](http://refreshless.com/wnumb/) is a number and money formatting library. Can be used independently, or with noUiSlider to format labels.
+A completed interactive requires the following to be included into the repository for release:
+
+- The interactive is primarily written in JS, HTML, and CSS.
```
Comment (JackMorganNZ) [inline docs/source/developer/interactives.rst]: Why is this a requirement?

File: docs/source/developer/interactives.rst [hunk 16]
```diff
@@ -1,126 +1,163 @@
...
+- The main file of the interactive must be ``index.html``.
+- A ``README.md`` file explaining the interactive, linking to any necessary wiki pages.
+- Extra libraries/scripts that have been used have been used and attributed correctly, see [details about how to do this here](CONTRIBUTING.md#managing-licenses).
+- Be easily accessible on desktop and mobile, or show a disclaimer that it is suited for desktop use only.
+- Abides by the repository contribution style guidelines.
+- All internal links (links to other files in the project) must be relative links.
```
Comment (JackMorganNZ) [inline docs/source/developer/interactives.rst]: This is out of date due to Django template tags.

File: csfieldguide/static/interactives/high-score-boxes/js/script.js [hunk 17]
```diff
@@ -110,7 +110,7 @@ function createBoxElements() {
 		box_image.id = 'box' + i;
 		box_image.className = 'box';
         box_image.draggable = false;
-		box_image.src = '../../img/colourful-box-images/square' + (box_images[i] + 1) + '-200px-rounded.png';
+		box_image.src = colourful_box_images[i];
```
Comment (JackMorganNZ) [inline csfieldguide/static/interactives/high-score-boxes/js/script.js]: Could you explain this?

File: csfieldguide/templates/interactives/iframe-interactive-base.html [hunk 18]
```diff
@@ -0,0 +1,22 @@
+{% load static %}
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/iframe-interactive-base.html]: Why isn't this template inheriting from the base template?

File: csfieldguide/templates/interactives/sorting-algorithm-comparison.html [hunk 19]
```diff
@@ -19,7 +19,7 @@
   <img id="sorting-algorithm-comparison-selection-sort-animation" class="sorting-algorithm-comparison-preload" src="{% static 'interactives/sorting-algorithm-comparison/img/selection-sort.gif' %}">
   <img id="sorting-algorithm-comparison-quick-sort-animation" class="sorting-algorithm-comparison-preload" src="{% static 'interactives/sorting-algorithm-comparison/img/quick-sort.gif' %}">
   
-  <link rel="stylesheet" href="{% static 'interactives/sorting-algorithm-comparison/css/sorting-algorithm-comparison.css' %}">
-  <script src="{% static 'interactives/sorting-algorithm-comparison/js/sorting-algorithm-comparison.js' %}"></script>
+  <link rel="stylesheet" href="{% static 'interactives/sorting-algorithm-comparison/css/style.css' %}">
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/sorting-algorithm-comparison.html]: Why are these files renamed?

File: csfieldguide/templates/interactives/whole-page-interactive-base.html [hunk 20]
```diff
@@ -1,22 +1,75 @@
 {% load static %}
 {% load render_html_field %}
 
+
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/whole-page-interactive-base.html]: Extra newline?

File: csfieldguide/templates/interactives/whole-page-interactive-base.html [hunk 21]
```diff
@@ -1,22 +1,75 @@
...
 <!DOCTYPE html>
 <html lang="en">
   <head>
-  
     <!-- Required meta tags -->
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+    <title>{{ interactive.name }} - CS Field Guide</title>
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/whole-page-interactive-base.html]: Should use "Computer Science Field Guide"

File: csfieldguide/templates/interactives/whole-page-interactive-base.html [hunk 22]
```diff
@@ -1,22 +1,75 @@
...
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+    <link rel="stylesheet" href="{% static 'css/website.css' %}">
+    <link rel="stylesheet" href="{% static 'css/third-party/featherlight.min.css' %}">
+    <link rel="stylesheet" href="{% static 'css/third-party/normalize.css' %}">
 
-    <title></title>
+    {% with 'interactives/'|add:interactive.slug|add:'/css/style.css' as css_static %}
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/whole-page-interactive-base.html]: Why is this here? Especially within the `<head>`

File: csfieldguide/templates/interactives/whole-page-interactive-base.html [hunk 23]
```diff
@@ -1,22 +1,75 @@
...
+              <img src="{% static 'img/logos/guide-logo-light.png' %}">
+              <a href="/en/" class="brand-logo">CSFG</a>
+              <h4>{{ interactive.name }}</h4>
+            </div>
+          </nav>
+        </div>
+
+      </header>
+
+    <main class="whole-page-interactive-content">
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/whole-page-interactive-base.html]: Why is the `body` wrapped in this?

File: csfieldguide/templates/interactives/whole-page-interactive-base.html [hunk 24]
```diff
@@ -1,22 +1,75 @@
...
+
+      <body>
+        <div class="row">
+          <div class="col s12 m10 l8 offset-l2 offset-m1">
+            <div class="interactive" id="interactive-high-score-boxes">
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/whole-page-interactive-base.html]: This `id` should be removed.

File: csfieldguide/templates/interactives/whole-page-interactive-base.html [hunk 25]
```diff
@@ -1,22 +1,75 @@
...
+              {% include interactive.template %}
+            </div>
+          </div>
+        </div>
+      </body>
+
+    </main>
+
+    {% with 'interactives/'|add:interactive.slug|add:'/js/script.js' as js_static %}
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/whole-page-interactive-base.html]: What does this achieve? I am still lost on it's purpose.

File: csfieldguide/templates/interactives/whole-page-interactive-base.html [hunk 26]
```diff
@@ -1,22 +1,75 @@
...
+      <script type="text/javascript" src="{% static js_static %}"></script>
+    {% endwith %}
+
+    <footer class="whole-page-interactive-footer">
+      <div class="footer-third">
+        <div class="container">
+          <a href="further-information/releases.html" class="version-number">v2.9.1</a>
+          <p>The Computer Science Field Guide uses a <a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons (CC BY-NC-SA 4.0) license</a>.</p>
```
Comment (JackMorganNZ) [inline csfieldguide/templates/interactives/whole-page-interactive-base.html]: It would be good to also reference our MIT licence here.

File: csfieldguide/chapters/content/en/chapters/coding-encryption.md [hunk 27]
```diff
@@ -0,0 +1,1156 @@
...
+It is very important that cryptosystems used in practice are semantically secure.
+
+As we saw above, Caesar Cipher is not semantically secure.
+
+{panel end}
+
+There are many other ways of cracking Caeser cipher which we will look at in this section.
+Understanding various common attacks on ciphers is important when looking at sophisticated cryptosystems which are used in practice.
+
+#### Frequency Analysis Attacks
```
[the remark this answers is not in the document]
Comment (hayleyavw) [reply csfieldguide/chapters/content/en/chapters/coding-encryption.md]: The headings used in csfg seem to be slightly inconsistent, most seem to capitalise key words (with many exceptions). I don't know what is meant to be our standard and I think this should be a separate issue, to establish a writing style guide that makes sense for csfg, and implement any changes required as a result.

[the remark this answers is not in the document]
Comment (hayleyavw) [reply csfieldguide/chapters/content/en/chapters/coding-encryption.md]: https://github.com/uccser/cs-field-guide/issues/578

File: csfieldguide/chapters/content/en/chapters/coding-encryption.md [hunk 28]
```diff
@@ -0,0 +1,1156 @@
...
+However, while this makes the Vigenere cipher more challenging to crack than the Caeser cipher, ways have been found to crack it quickly.
+In fact, once you know the key length, it just breaks down to cracking several Caesar ciphers (which as you have seen is straightforward, and you can even use frequency analysis on the individual Caesar Ciphers!) Several statistical methods have been devised for working out the key length.
+
+Attacking the Vigenere cipher by trying every possible key is hard because there are a lot more possible keys than for the Caesar cipher, but a statistical attack can work quite quickly.
+The Vigenere cipher is known as a *polyalphabetic substitution cipher*, since it is uses multiple substitution rules.
+
+{panel end}
+
+
+#### Known Plain Text Attacks
```
[the remark this answers is not in the document]
Comment (hayleyavw) [reply csfieldguide/chapters/content/en/chapters/coding-encryption.md]: see #578

File: csfieldguide/chapters/content/en/chapters/coding-encryption.md [hunk 29]
```diff
@@ -0,0 +1,1156 @@
...
+
+{comment interactive name="rsa-jsencrypt" type="whole-page" text="RSA Decrypter (using padding)" parameters="mode=decrypt"}
+
+{comment #### How does RSA Work?}
+
+{comment #### Preventing known plain text attacks}
+
+{comment #### Common usage}
+
+## Storing Passwords Securely
```
[the remark this answers is not in the document]
Comment (hayleyavw) [reply csfieldguide/chapters/content/en/chapters/coding-encryption.md]: see #578

File: csfieldguide/chapters/content/en/chapters/coding-encryption.md [hunk 30]
```diff
@@ -0,0 +1,1156 @@
...
+Another major problem is writing secure software that doesn't leave ways for a user to get access to information that they shouldn't (such as typing a database command into a website query and have the system accidentally run it, or overflowing the buffer with a long input, which could accidentally replace parts of the program).
+Also, systems need to be protected from "denial of service" (DOS) attacks, where they get so overloaded with requests (e.g. to view a web site) that the server can't cope, and legitimate users get very slow response from the system, or it might even fail completely.
+
+For other kinds of attacks relating to computer security, see the [Wikipedia entry on Hackers](https://en.wikipedia.org/wiki/Hacker_(computer_security)).
+
+There's a dark cloud hanging over the security of all current encryption methods: [Quantum computing](https://en.wikipedia.org/wiki/Quantum_computer).
+Quantum computing is in its infancy, but if this approach to computing is successful, it has the potential to run very fast algorithms for attacking our most secure encryption systems (for example, it could be used to factorise numbers very quickly).
+In fact, the quantum
```
[the remark this answers is not in the document]
Comment (hayleyavw) [reply csfieldguide/chapters/content/en/chapters/coding-encryption.md]: https://github.com/uccser/cs-field-guide/issues/578

File: csfieldguide/chapters/management/commands/_ChapterLoader.py [hunk 31]
```diff
@@ -45,23 +53,43 @@ def load(self):
         if chapter_number is None:
             raise MissingRequiredFieldError(
                 self.structure_file_path,
-                ["number"],
-                "Chapter Number"
+                self.chapter_required_fields,
+                "Chapter"
             )
 
         chapter_icon = self.chapter_structure.get('icon', None)
```
[the remark this answers is not in the document]
Comment (hayleyavw) [reply csfieldguide/chapters/management/commands/_ChapterLoader.py]: following the convention we set in csu where we are always pretty explicit

File: csfieldguide/config/settings/base.py (hunk 2 above)
Comment (hayleyavw) [reply csfieldguide/config/settings/base.py]: I've been struggling with this bit, removing either line causes different errors, so I realise there is some overlap between them, but I can't work out why each option is causing different issues. Would be helpful if you could spin it up and have a tinker and see if you can spot what has happened to cause this.

File: docs/source/developer/interactives.rst (hunk 15 above)
Comment (hayleyavw) [reply docs/source/developer/interactives.rst]: this was copied from the original interactive docs, and is probably in need of a discussion! have removed it for now.

File: csfieldguide/templates/interactives/sorting-algorithm-comparison.html (hunk 19 above)
Comment (hayleyavw) [reply csfieldguide/templates/interactives/sorting-algorithm-comparison.html]: consistency, there isn't really a standard amongst out interactive filenames atm.
Makes just as much sense to call everything `<interactive-slug>.html`/`css`/`js` etc, just `style` and `script` is a convention I've seen a lot. Up for discussion.

File: csfieldguide/templates/interactives/iframe-interactive-base.html (hunk 18 above)
Comment (hayleyavw) [reply csfieldguide/templates/interactives/iframe-interactive-base.html]: not sure that I understand why it would need to?

**Comment by hayleyavw** [comment]
@JackMorganNZ 👋  ready for re-review.

**Comment by JackMorganNZ** [comment]
Adding comment now as I'm having a brainwave on my annual leave, will think it through and build upon it next week.

> It may be more elegant to include CSS files in YAML to save on model, and then store relationship to chapter/section (need to figure that out) to retrieve required CSS files for <head> of specific page.

**Comment by JackMorganNZ** [comment]
Dismissing my own review as I'm working on this PR now instead of @hayleyavw.

**Comment by hayleyavw** [comment]
LGTM, but, the static files problem is back (see screenshot attached).....
screen shot 2018-03-18 at 9 48 56 pm
